fennel/client_tests/branches/test_clone.py

```diff
@@ -0,0 +1,390 @@
+    """
+    client.sync(
+        datasets=[UserInfoDataset, GenderStats],
+    )
+    client.clone_branch("test-branch", from_branch="main")
+
+    now = datetime.now()
+    data = [
+        {
+            "user_id": 1,
+            "name": "John",
+            "age": 30,
+            "gender": "male",
+            "country_code": 1,
+            "email": "john@fennel",
+            "timestamp": now,
+        },
+        {
+            "user_id": 2,
+            "name": "Rachel",
+            "age": 55,
+            "gender": "F",
+            "country_code": 1,
+            "email": "rachel@fennel",
+            "timestamp": now,
+        },
+    ]
+    df = pd.DataFrame(data)
+    response = client.log("fennel_webhook", "UserInfoDataset", df)
+    assert response.status_code == requests.codes.OK, response.json()
+
+    assert client.branch == "test-branch"
+    output = client.get_dataset_df(
+        "GenderStats",
+    )
+    assert output.shape == (2, 3)
+
+    client.checkout("main")
+    assert client.branch == "main"
+    output = client.get_dataset_df("GenderStats")
+    assert output.shape == (2, 3)
+
+
+@pytest.mark.integration
+@mock
+def test_add_dataset_clone_branch(client):
+    """
+    Cling a branch, then adding one or more datasets in cloned branch. Change should be reflected in cloned branch only.
+    """
+    client.sync(
+        datasets=[UserInfoDataset, GenderStats],
+    )
+    client.clone_branch("test-branch", from_branch="main")
+
+    client.sync(
+        datasets=[UserInfoDataset, GenderStats, CountryStats],
+    )
+
+    now = datetime.now()
+    data = [
+        {
+            "user_id": 1,
+            "name": "John",
+            "age": 30,
+            "gender": "male",
+            "country_code": 1,
+            "email": "john@fennel",
+            "timestamp": now,
+        },
+        {
+            "user_id": 2,
+            "name": "Rachel",
+            "age": 55,
+            "gender": "F",
+            "country_code": 2,
+            "email": "rachel@fennel",
+            "timestamp": now,
+        },
+    ]
+    df = pd.DataFrame(data)
+    response = client.log("fennel_webhook", "UserInfoDataset", df)
+    assert response.status_code == requests.codes.OK, response.json()
+
+    output = client.get_dataset_df("GenderStats")
+    assert output.shape == (2, 3)
+
+    output = client.get_dataset_df("CountryStats")
+    assert output.shape == (2, 3)
+
+    client.checkout("main")
+    output = client.get_dataset_df("GenderStats")
+    assert output.shape == (2, 3)
+
+    with pytest.raises(ValueError) as error:
+        client.get_dataset_df("CountryStats")
+    assert str(error.value) == "Dataset `CountryStats` not found"
+
+
+@pytest.mark.integration
+@mock
+def test_change_dataset_clone_branch(client):
+    """
+    Clone a branch A → B. Verify A & B both give the same answers.
+    Then modify A. Ensure B keeps giving the same answers.
+    """
+    now = datetime.now()
```

Comment:
Can you add a test where we have branch A, with src -> pipeline.
Clone A -> B. 
In B we modify the source dataset itself. ( different webhook ) now the derived pipelines give different answers

Reply:
added